v3/lints/cabf_br/lint_ca_invalid_eku.go

```diff
@@ -0,0 +1,81 @@
+/*
+ * ZLint Copyright 2024 Regents of the University of Michigan
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+/*
+ * Contributed by Adriano Santoni <asantoni64@gmail.com>
+ */
+
+package cabf_br
+
+import (
+	"github.com/zmap/zcrypto/x509"
+	"github.com/zmap/zlint/v3/lint"
+	"github.com/zmap/zlint/v3/util"
+
+	"fmt"
+)
+
+func init() {
+	lint.RegisterCertificateLint(&lint.CertificateLint{
+		LintMetadata: lint.LintMetadata{
+			Name:          "e_ca_invalid_eku",
+			Description:   "Checks that SubCA certificates do not contain forbidden values in their EKU extension.",
+			Citation:      "CABF BRs §7.1.2",
+			Source:        lint.CABFBaselineRequirements,
+			EffectiveDate: util.CABFBRs_1_7_1_Date,
+		},
+		Lint: NewCaInvalidEKU,
+	})
+}
+
+type caInvalidEKU struct{}
+
+func NewCaInvalidEKU() lint.LintInterface {
+	return &caInvalidEKU{}
+}
+
+func (l *caInvalidEKU) CheckApplies(c *x509.Certificate) bool {
+	// This lint applies to any SubCA certificate to which the CABF BRs are applicable and which contains
+	// the EKU extension. Given that the lint source is lint.CABFBaselineRequirements, if we arrive here
+	// it's been already checked that the certificate falls within the purview of the CABF BRs.
+	return util.IsSubCA(c) && len(c.ExtKeyUsage) != 0
+}
+
+func (l *caInvalidEKU) Execute(c *x509.Certificate) *lint.LintResult {
+
+	// If the EKU contains anyExtendedKeyUsage, it's probably a cross-certicate
+	// In this case, the EKU must not contain any other value
+	if util.HasEKU(c, x509.ExtKeyUsageAny) && len(c.ExtKeyUsage) > 1 {
+		return &lint.LintResult{
+			Status:  lint.Error,
+			Details: "anyExtendedKeyUsage MUST NOT be accompanied by any other value in the EKU extension",
+		}
+	}
+
+	// If we get here, it is necessarily a SubCA with serverAuth in the EKU
+	for _, eku := range c.ExtKeyUsage {
+		if eku == x509.ExtKeyUsageEmailProtection ||
+			eku == x509.ExtKeyUsageCodeSigning ||
+			eku == x509.ExtKeyUsageTimeStamping ||
+			eku == x509.ExtKeyUsageOcspSigning {
```

Comment:
I believe that we are referring to sub-section g?

> For Subordinate CA Certificates that will be used to issue TLS certificates, the value
id-kp-serverAuth [RFC5280] MUST be present. The value id-kp-clientAuth
[RFC5280] MAY be present. The values id-kp-emailProtection [RFC5280], id-kpcodeSigning [RFC5280], id-kp-timeStamping [RFC5280], and anyExtendedKeyUsage
[RFC5280] MUST NOT be present. Other values SHOULD NOT be present.
>
>For Subordinate CA Certificates that are not used to issue TLS certificates, then the
value id-kp-serverAuth [RFC5280] MUST NOT be present. Other values MAY be
present, but SHOULD NOT combine multiple independent usages (e.g. including idkp-timeStamping [RFC5280] with id-kp-codeSigning [RFC5280]).

So unless my eyes are missing it, I think that since OCSP Signing is not explicitly enumerated that it would fall under the `Other values SHOULD NOT be present` clause, thus making it a warning?

Reply:
> So unless my eyes are missing it, I think that since OCSP Signing is not explicitly enumerated that it, so would fall under the `Other values SHOULD NOT be present` clause, thus making it a warning?

This list matches [7.1.2.10.6 CA Certificate Extended Key Usage](https://github.com/cabforum/servercert/blob/049237e096650fe01f67780b7c24bd5211ee3038/docs/BR.md#712106-ca-certificate-extended-key-usage).

Reply:
@mathewhodson you are correct, it is enumerated in latest. This is where it gets mildly tricky since this lint is marked as being effective as of August 2020 via [CABF/BR 1.7.1](https://cabforum.org/uploads/CA-Browser-Forum-BR-1.7.1.pdf) where I don't think I see it marked as a `MUST NOT`.

@defacto64 did you have a particular reason for choosing 1.7.1?